clients/cobol-lsp-vscode-extension/src/services/CopybookURI.ts

```diff
@@ -0,0 +1,119 @@
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+import * as fs from "fs";
+import * as path from "path";
+import {URL} from "url";
+import * as vscode from "vscode";
+import {C4Z_FOLDER, COPYBOOKS_FOLDER, PATHS_LOCAL_KEY, PATHS_ZOWE, SETTINGS_SECTION} from "../constants";
+import {CopybooksDownloader} from "./CopybooksDownloader";
+import {ProfileService} from "./ProfileService";
+import {CopybookResolver} from "./settings/CopybookResolver";
+import {LocalCopybookResolver} from "./settings/LocalCopybookResolver";
+import {CopybookUtil} from "./settings/util/CopybookUtil";
+
+/**
+ * This class is responsible to identify from which source resolve copybooks required by the server.
+ */
+export class CopybookURI {
+    /**
+     * This method scan the list of folders as given input and find the required copybook name within the folder.
+     * If found returns its URI representation
+     * @param copybookName name of the copybook asked by the server
+     * @param targetFolders list of folders from where to search the copybook
+     */
+    public static searchInWorkspace(copybookName: string, targetFolders: string[]): string {
+        const copybookResolver: CopybookResolver = new LocalCopybookResolver();
+        const localFolderList: string[] = copybookResolver.resolve(targetFolders);
+
+        for (const folder of localFolderList) {
+            let uri: URL = this.getURIFrom(folder, copybookName);
+
+            if (fs.existsSync(uri)) {
+                return uri.href;
+            } else {
+                uri = this.getURIFrom(folder, copybookName, CopybookUtil.getValidCopybookExtensionList());
+                if (fs.existsSync(uri)) {
+                    return uri.href;
+                }
+            }
+        }
+    }
+
+    /**
+     * This method is responsible to return an URI without extension if the extension is not provided or an URI
+     * that contains an allowed extension.
+     * @param folder is the first part of the URI referred to the folder defined in the setting.json
+     * @param copybookName is the name of copybook identified by the LSP server that needs to be found locally
+     * or downloaded
+     * @param extensions an optional parameter to produce an URI of an allowed extension list, verifyng that
+     * this URI really exists on FS.
+     */
+    private static getURIFrom(folder: string, copybookName: string, extensions?: string[]): URL {
+        if (!extensions) {
+            return new URL(path.join(folder, copybookName));
+        } else {
+            for (const extension of extensions) {
+                const uri = new URL(path.join(folder, copybookName + "." + extension));
+                if (fs.existsSync(uri)) {
```

Comment:
In case when we have many extensions we will do many calls to FS to check the file existing state. Things will go harder in a case when we also want to support case insensitive scenario.
I think what is possible to take a list of files from the folder once and then work with that list.

Reply:
> In case when we have many extensions we will do many calls to FS to check the file existing state. Things will go harder in a case when we also want to support case insensitive scenario.
> I think what is possible to take a list of files from the folder once and then work with that list.

Well.. to be honest I am not expecting to have more than 3 elements (cpy, CPY, Cpy) so I don't really see the value to create a more expensive operation

> Also, this check is case-sensitive
True - I will fix this immediately :)

Reply:
I will fix it